crates/circuits/sha256-air/src/sha256/trace.rs

```diff
@@ -0,0 +1,405 @@
+        local_cols: &Sha256RoundCols<F>,
+        next_cols: &mut Sha256RoundCols<F>,
+        row_idx: usize,
+    ) {
+        let a = [local_cols.work_vars.a, next_cols.work_vars.a].concat();
+        let e = [local_cols.work_vars.e, next_cols.work_vars.e].concat();
+        for i in 0..SHA256_ROUNDS_PER_ROW {
+            let cur_a = a[i + 4];
+            let sig_a = big_sig0_field::<F>(&a[i + 3]);
+            let maj_abc = maj_field::<F>(&a[i + 3], &a[i + 2], &a[i + 1]);
+            let d = a[i];
+            let cur_e = e[i + 4];
+            let sig_e = big_sig1_field::<F>(&e[i + 3]);
+            let ch_efg = ch_field::<F>(&e[i + 3], &e[i + 2], &e[i + 1]);
+            let h = e[i];
+            let w = if row_idx < 16 {
+                next_cols.message_schedule.w[i]
+            } else {
+                [F::ZERO; 32]
+            };
+
+            let t1 = [h, sig_e, ch_efg, w];
+            let t2 = [sig_a, maj_abc];
+            for j in 0..SHA256_WORD_U16S {
+                let t1_limb_sum = t1.iter().fold(F::ZERO, |acc, x| {
+                    acc + compose::<F>(&x[j * 16..(j + 1) * 16], 1)
+                });
+                let t2_limb_sum = t2.iter().fold(F::ZERO, |acc, x| {
+                    acc + compose::<F>(&x[j * 16..(j + 1) * 16], 1)
+                });
+                let d_limb = compose::<F>(&d[j * 16..(j + 1) * 16], 1);
+                let k_limb = if row_idx < 16 {
+                    F::from_canonical_u32(
+                        u32_into_limbs::<SHA256_WORD_U16S>(
+                            SHA256_K[row_idx * SHA256_ROUNDS_PER_ROW + i],
+                        )[j],
+                    )
+                } else {
+                    F::ZERO
+                };
+                let cur_a_limb = compose::<F>(&cur_a[j * 16..(j + 1) * 16], 1);
+                let cur_e_limb = compose::<F>(&cur_e[j * 16..(j + 1) * 16], 1);
+                let sum = d_limb
+                    + t1_limb_sum
+                    + k_limb
+                    + if j == 0 {
+                        F::ZERO
+                    } else {
+                        next_cols.work_vars.carry_e[i][j - 1]
+                    }
+                    - cur_e_limb;
+                let carry_e = sum * (F::from_canonical_u32(1 << 16).inverse());
+
+                let sum = t1_limb_sum
+                    + t2_limb_sum
+                    + k_limb
+                    + if j == 0 {
+                        F::ZERO
+                    } else {
+                        next_cols.work_vars.carry_a[i][j - 1]
+                    }
+                    - cur_a_limb;
+                let carry_a = sum * (F::from_canonical_u32(1 << 16).inverse());
+                next_cols.work_vars.carry_e[i][j] = carry_e;
+                next_cols.work_vars.carry_a[i][j] = carry_a;
+            }
+        }
+    }
+
+    /// Puts the correct intermed_4 in the `next_row`
+    pub fn generate_intermed_4<F: PrimeField32>(
```

Comment:
this is not called in `generate_block_trace`! that makes the API overall hard to use and error-prone. also impacts parallelism.